Category CRUD + 모든 카테고리 기능 생성

DeiLee0913 opened the pull request:
## 📝 작업 내용

- [x]  카테고리 생성 API
- [x]  ID로 카테고리 수정 API
- [x]  NAME으로 카테고리 수정 API
- [x]  카테고리 조회 API
- [x]  ID로 카테고리 삭제 API
- [x]  NAME으로 카테고리 삭제 API
- [x]  캘린더 ID로 모든 카테고리 조회 API

## 📸 결과

- ‘POST’ 카테고리 생성(/api/v1/category)
    - RequestBody
        
        ```json
        {
            "calendarId": "3",
            "name": "Dev"
        }
        ```
        
    - ReponseBody
        
        ```json
        {
            "categoryId": 11,
            "calendarId": 3,
            "name": "Dev"
        }
        ```
        
- ‘GET’ 카테고리 조회(/api/v1/category/{categoryId})
    - ResponseBody
        
        ```kotlin
        {
            "categoryId": 11,
            "calendarId": 3,
            "name": "Dev"
        }
        ```
        
- ‘PUT’ ID로 카테고리 수정(/api/v1/category/by-id)
    - RequestBody
        
        ```json
        {
            "categoryId": "1",
            "newName": "Passive Learning"
        }
        ```
        
    - RespnseBody
        
        ```json
        {
            "categoryId": 1,
            "calendarId": 3,
            "name": "Passive Learning"
        }
        ```
        
- ‘PUT’ 이름으로 카테고리 수정(/api/v1/category/by-name)
    - RequestBody
        
        ```json
        {
            "originalName": "Archiving",
            "newName": "Idea"
        }
        ```
        
    - RespnseBody
        
        ```json
        {
            "categoryId": 4,
            "calendarId": 3,
            "name": "Idea&Archive"
        }
        ```
        
- ‘DELETE’ ID로 카테고리 삭제(/api/v1/category/by-id/{calendarId})
- ‘DELETE’ NAME으로 카테고리 삭제(/api/v1/category/by-name/{calendarName})

## 💬 리뷰 요구사항 (선택)

## #️⃣ 연관된 이슈 (선택)

DeiLee0913 committed:
✨ feat: Category Domain, Dto

DeiLee0913 committed:
✨ feat: Category CRUD + getAllCategories

- Category CRUD 구현
: create, upate 시에는 이미 같은 이름으로 존재하는 category가 있는지 확인
- getAllCategories 구현
: calendarId로 해당 캘린더의 모든 Category 받아오기

DeiLee0913 committed:
✨update: Category updateByName

- 기존 update를 updateById로 수정하고 updateByName 추가

DeiLee0913 committed:
✨update: Category deleteByName

chlqls commented on src/main/java/jusomejusome/togather/category/controller/CategoryController.java:
```diff
@@ -0,0 +1,66 @@
+
+    @PostMapping
+    public ResponseEntity<CategoryResDto> createCategory(@RequestBody CategoryCreateReqDto createDto) {
+        CategoryResDto createdCateogry = categoryService.createCategory(createDto);
```
아주 사소한건데 `createdCateogry` 여기 오타 났어여!

chlqls commented on src/main/java/jusomejusome/togather/category/controller/CategoryController.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    @DeleteMapping("/by-id/{categoryId}")
+    public ResponseEntity<Void> deleteCategoryById(@PathVariable Long categoryId) {
```
`deleteCategoryByName`도 마찬가지인데, 카테고리 삭제 시 반환 값이 없으면 제대로 삭제되었는지 확인하기 어려우니까 "SUCCESS" 또는 "삭제되었습니다" 등의 간단한 **반환 값**이 있는게 좋을 것 같아요!

chlqls commented on src/main/java/jusomejusome/togather/category/service/CategoryService.java:
```diff
@@ -0,0 +1,114 @@
+
+    @Transactional
+    public CategoryResDto createCategory(CategoryCreateReqDto createDto) {
+        //동일한 이름을 가진 카테고리가 있는지 확인
+        if(categoryRepository.existsByName(createDto.getName())) {
+            throw new CustomException(ErrorCode.CATEGORY_EXISTED_EXCEPTION);
+        }
```
같은 이름의 카테고리더라도 캘린더가 다를 수 있으니까 여기서 **캘린더 id**도 같은지 확인해줘야할 것 같아요! 카데고리 업데이트할 때도 마찬가지로 수정하면 될 것 같습니댜

DeiLee0913 committed:
✨update: 검증 로직 구현, 이름으로 작업시 API 변수 타입 변경

- 동일 캘린더 안에서만 카테고리 중복을 검사하도록 변경
- 카테고리에 대한 유저의 접근 권한을 검사하는 로직 추가
- 카테고리 이름으로 카테고리를 찾거나 존재 여부를 확인할 때 캘린더 id도 함께 받아 해당 캘린더 내에서 카테고리를 찾도록 함
- 카테고리 이름을 받는 경우 PathVariable이 아닌 NameReqDto를 사용하도록 변경
